### ISSUE TITLE
bug: zellij: It is not clear how to configure keyboard bindings

### ISSUE THREAD

**Issue by brianmay** [nix-community/home-manager#4659]
### Are you following the right branch?

- [X] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

From the docs zellij keyboard bindings look like:

```
    // bind the Alt-n to open a new pane
    bind "Alt n" { NewPane; }
    // bind both the "h" key and the left-arrow key to move pane focus left
    bind "h" "Left" { MoveFocus "Left"; }
    // bind the "f" key to toggle the focused pane full-screen and switch to normal mode
    bind "f" { ToggleFocusFullscreen; SwitchToMode "Normal"; }
```

But it is far from clear what I have to set the `settings` to to generate this. And I can't see any documentation either.

I have seen bug #4054, but this hasn't really helped.

### Maintainer CC

@mainr

### System information

```markdown
- system: `"x86_64-linux"`
 - host os: `Linux 6.1.61, NixOS, 23.05 (Stoat), 23.05.20231105.aeefe20`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.6`
 - channels(brian): `""`
 - channels(root): `"nixos-23.05"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by brianmay** [comment]
After writing this I found #4160, looks like there isn't a solution yet. But when the solution is found, it should be clearly documented...

**Comment by brianmay** [comment]
Oh, that explains my confusion. I wasn't aware zellij changed formats.

**Comment by pluiedev** [comment]
Yeah they did, in [version 0.32.0](https://github.com/zellij-org/zellij/releases/tag/v0.32.0), although we should probably document this in the module somewhere as it can break things, and users should be aware of that

**Comment by gotcha** [comment]
I have added documentation: https://github.com/nix-community/home-manager/pull/4665

Feedback welcome !

**Comment by brianmay** [comment]
I think this is still a concern. There is a proposal in #4465 to resolve, but looks like that has stalled too :-(.

**Comment by b-m-f** [comment]
Still an issue

**Comment by pluiedev** [comment]
Still a thing.

**Comment by Dietr1ch** [comment]
Can we get a textual `extraSettings` out to allow configuring things?

I guess that figuring out the best way to transform nix values into kdl configs in a more generic way is going to take a while.

**Comment by brianmay** [comment]
Have a look at the KDL spec: https://kdl.dev/

I think though the terminating semicolon may not be important. when there is only one node. Or nodes a separated by new lines.

But what is wrong, is we have to represent a list of nodes with a attrset.

```nix
{
   SwitchToMode = "Normal";
};
```

But it isn't a attrset...

* We don't have unique keys. `SwitchToMode` is not a unique key, we can repeat that numerous times.
* Order of nodes is important, but we can't preserve the order in a attrset.
* Nodes may have a name but no arguments, something we cannot represent.

An example of a valid KDL document is:

```kdl
contents {
  section "First section" {
    paragraph "This is the first paragraph"
    paragraph "This is the second paragraph"
  }
}
```

A array for list of nodes would be better.

I suspect what is happened is that `toKDL` was built without referencing the KDL spec. And as a result it isn't really a good generic solution to writing KDL files.

On the positive side though, I think zellij is the only program that uses `toKDL`, so is the only program that we risk breaking if we try to fix this.

My feeling is we should support something like:

```nix
keybinds = {
       locked = [
            {
              name = "bind";
              args=["Ctrl h"];
              children={ name="SwitchToMode"; args=["Normal"]; };
            }
          ] 
       ];
};
```

(apologies if there are syntax errors here...)

Not sure if children have to appear last or if they can appear anywhere in the list.

I think this would be sufficient for zellij, although I still haven't tried to support properties:

```kdl
author "Alex Monad" email="alex@example.com" active=true
```

Or type annotations:

```kdl
numbers (u8)10 (i32)20 myfloat=(f32)1.5 {
  strings (uuid)"123e4567-e89b-12d3-a456-426614174000" (date)"2021-02-03" filter=(regex)r"$\d+"
  (author)person name="Alex"
}
```

**Comment by Dietr1ch** [comment]
> I suspect what is happened is that `toKDL` was built without referencing the KDL spec. And as a result it isn't really a good generic solution to writing KDL files.

To be fair their websites call their [https://github.com/kdl-org/kdl/blob/main/SPEC.md](Specification) a living one, and while it's now at version `1.0.0` I'm guessing `toKDL` might precede it.

> On the positive side though, I think zellij is the only program that uses `toKDL`, so is the only program that we risk breaking if we try to fix this.

I believe so too. KDL's [FAQ](https://kdl.dev#faq) asks "Why yet another document language?" and "Have you seen that one XKCD comic about standards?", but I guess that comic #927 wasn't strong enough to avoid having another language around that just introduces new syntax without expressiveness worth its weight (tooling, debugger, libraries, it's expensive to have your own language).

Beyond what I believe, you seem to be right too, here's a [search for KDL on the repo](https://github.com/search?q=repo%3Anix-community%2Fhome-manager+kdl&type=code), it only mentions the new files to handle KDL and [home-manager/modules/programs/zellij.nix](https://github.com/nix-community/home-manager/tree/master/modules/programs/zellij.nix) is the only user of all the KDL machinery.

**Comment by Fymyte** [comment]
In case someone gets here, I managed to make it *almost* work using ideas from https://github.com/nix-community/home-manager/issues/4659#issuecomment-2470388666

```nix
{
  lib,
  ...
}: let
  zellijBindings = bindList:
    lib.attrsets.mergeAttrsList (lib.lists.forEach bindList (el: {"bind \"${el.key}\"" = el.action;}));
in {
  programs.zellij.enable = true;
  programs.zellij = {
    settings = {
      keybinds = {
        normal = zellijBindings [
          {
            key = "Ctrl f";
            action = {SwitchToMode = "Tmux";};
          }
          {
            key = "Ctrl a";
            action = {SwitchToMode = "Pane";};
          }
        ];
      };
    };
  };
}
```
There is still some gotcha in case you want to remove default keybindings, for example:
```nix
      keybinds = {
        "normal clear-defaults=true" = zellijBindings [
          {
            key = "Ctrl f";
            action = {SwitchToMode = "Tmux";};
          }
        ];
      };
```
and if you want to unbind:
```nix
        "normal clear-defaults=true" = zellijBindings [
          {
            key = "Ctrl f";
            action = {SwitchToMode = "Tmux";};
          }
        ] // {unbind = "b";};
      };
```

**Comment by pluiedev** [comment]
Stale bot is the scourge of the earth

**Comment by jpentland** [comment]
For keybinds it seems like maybe a list is more appropriate than a set? Since order matters and values can be repeated. So the module could have an extra "keybinds" option that generates this part of the config from a list , e.g.

```
programs.zellij.keybinds.locked = [
    { bind.c.NewPane = true; } # For no args true seems more idiomatic, null implies perhaps that something is being disable or cancelled out
    { bind.u.SwitchToMode = "normal"; }
];
```
